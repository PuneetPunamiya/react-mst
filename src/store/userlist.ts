import { types, Instance } from "mobx-state-tree";
import { values } from "mobx";

export const User = types.model({
  name: types.optional(types.string, ""),
  age: types.optional(types.integer, 0),
  class: types.optional(types.string, ""),
});

export const Store = types
  .model({
    users: types.array(User),
    filter: types.optional(types.array(types.string), []),
    sort: types.optional(types.string, ""),
  })
  .views((self) => ({
    get count() {
      return self.users.length;
    },
    get userList() {
      return filterAndSortUsers(values(self.users), self.filter, self.sort);
    },
  }))
  .actions((self) => ({
    add(item: UserItem) {
      self.users.push(item);
    },
    setSort(sort: string) {
      self.sort = sort;
    },
    setFilter(filter: any) {
      self.filter = filter;
    },
  }));

export interface UserItem extends Instance<typeof User> {}

function filterAndSortUsers(users: any, filter: any, sort: any) {
  if (filter.length === 0 && sort === "") {
    return users;
  }
  if (filter.length === 0) {
    if (sort === "name") {
      return users.sort((a: UserItem, b: UserItem) =>
        a.name > b.name ? 1 : a.name === b.name ? 0 : -1
      );
    } else if (sort === "age") {
      return users.sort((a: UserItem, b: UserItem) =>
        a.age > b.age ? 1 : a.age === b.age ? 0 : -1
      );
    } else {
      return users;
    }
  }
  if (sort === "name") {
    return users
      .filter((u: UserItem) => filter.includes(u.class))
      .sort((a: UserItem, b: UserItem) =>
        a.name > b.name ? 1 : a.name === b.name ? 0 : -1
      );
  } else if (sort === "age") {
    return users
      .filter((u: UserItem) => filter.includes(u.class))
      .sort((a: UserItem, b: UserItem) =>
        a.age > b.age ? 1 : a.age === b.age ? 0 : -1
      );
  } else {
    return users.filter((u: UserItem) => filter.includes(u.class));
  }
}
