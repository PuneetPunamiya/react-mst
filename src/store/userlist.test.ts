import { getSnapshot, onSnapshot, onPatch } from "mobx-state-tree";
import { User, Store } from "./userlist";

it("can create an user", () => {
  const user = User.create({
    name: "test",
    age: 12,
    class: "SY",
  });

  expect(user.name).toBe("test");
  expect(user.age).toBe(12);
  expect(user.class).toBe("SY");
});

it("can create a store", () => {
  const userlist = Store.create({
    users: [
      { name: "test1", age: 12, class: "SY" },
      { name: "test2", age: 13, class: "TY" },
    ],
  });

  expect(userlist.users[0].name).toBe("test1");
  expect(userlist.users[1].name).toBe("test2");
});

it("can create empty store", () => {
  const list = Store.create({});
  expect(list.users.length).toBe(0);
});

it("can get number of users in a store", () => {
  const store = Store.create({
    users: [
      { name: "test1", age: 12, class: "SY" },
      { name: "test2", age: 13, class: "TY" },
    ],
  });
  const numberOfUser = store.count;
  expect(numberOfUser).toBe(2);
});

it("can add user to users list in Store", () => {
  const userStore = Store.create();

  const states = [];
  onSnapshot(userStore, (s) => states.push(s));
  const patches = [];
  onPatch(userStore, (p) => patches.push(p));

  userStore.add({ name: "test", age: 43, class: "SY" });
  const first = userStore.users[0];

  expect(first.name).toBe("test");
  expect(getSnapshot(userStore)).toMatchSnapshot();
  expect(patches).toMatchSnapshot();
});

it("can get users sorted by name", () => {
  const userStore = Store.create();

  const states = [];
  onSnapshot(userStore, (s) => states.push(s));
  const patches = [];
  onPatch(userStore, (p) => patches.push(p));

  userStore.add({ name: "rachel", age: 18, class: "FY" });
  userStore.add({ name: "joey", age: 27, class: "FY" });
  userStore.add({ name: "chandler", age: 28, class: "SY" });

  userStore.setSort("name");

  const list = userStore.userList
  expect(list[0].name).toBe("chandler");
  expect(getSnapshot(userStore)).toMatchSnapshot();
  expect(patches).toMatchSnapshot();
});

it("can get users sorted by age", () => {
  const userStore = Store.create();

  const states = [];
  onSnapshot(userStore, (s) => states.push(s));
  const patches = [];
  onPatch(userStore, (p) => patches.push(p));

  userStore.add({ name: "rachel", age: 18, class: "FY" });
  userStore.add({ name: "joey", age: 27, class: "FY" });
  userStore.add({ name: "chandler", age: 28, class: "SY" });

  userStore.setSort("age");

  const list = userStore.userList
  expect(list[0].name).toBe("rachel");
  expect(list[0].age).toBe(18);
  expect(getSnapshot(userStore)).toMatchSnapshot();
  expect(patches).toMatchSnapshot();
});

it("can get users filtered by class", () => {
  const userStore = Store.create();

  const states = [];
  onSnapshot(userStore, (s) => states.push(s));
  const patches = [];
  onPatch(userStore, (p) => patches.push(p));

  userStore.add({ name: "rachel", age: 18, class: "FY" });
  userStore.add({ name: "joey", age: 27, class: "FY" });
  userStore.add({ name: "chandler", age: 28, class: "SY" });

  userStore.setFilter(["FY"]);
  const list = userStore.userList
  expect(list.length).toBe(2);

  expect(getSnapshot(userStore)).toMatchSnapshot();
  expect(patches).toMatchSnapshot();
});

it("can get users filtered by class and sorted", () => {
  const userStore = Store.create();

  const states = [];
  onSnapshot(userStore, (s) => states.push(s));
  const patches = [];
  onPatch(userStore, (p) => patches.push(p));

  userStore.add({ name: "rachel", age: 18, class: "FY" });
  userStore.add({ name: "joey", age: 27, class: "FY" });
  userStore.add({ name: "chandler", age: 28, class: "SY" });

  userStore.setSort("name");
  userStore.setFilter(["FY"]);
  const list = userStore.userList
  expect(list.length).toBe(2);
  expect(list[0].name).toBe("joey");

  expect(getSnapshot(userStore)).toMatchSnapshot();
  expect(patches).toMatchSnapshot();
});
