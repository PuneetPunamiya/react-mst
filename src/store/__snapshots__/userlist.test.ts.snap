// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add user to users list in Store 1`] = `
Object {
  "filter": Array [],
  "sort": "",
  "users": Array [
    Object {
      "age": 43,
      "class": "SY",
      "name": "test",
    },
  ],
}
`;

exports[`can add user to users list in Store 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/0",
    "value": Object {
      "age": 43,
      "class": "SY",
      "name": "test",
    },
  },
]
`;

exports[`can add user to users list in Store 3`] = `
Object {
  "filter": Array [],
  "sort": "",
  "users": Array [
    Object {
      "age": 43,
      "class": "SY",
      "name": "test",
    },
  ],
}
`;

exports[`can add user to users list in Store 4`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/0",
    "value": Object {
      "age": 43,
      "class": "SY",
      "name": "test",
    },
  },
]
`;

exports[`can get users filtered by class 1`] = `
Object {
  "filter": Array [
    "FY",
  ],
  "sort": "",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
    Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
    Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  ],
}
`;

exports[`can get users filtered by class 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/0",
    "value": Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
  },
  Object {
    "op": "add",
    "path": "/users/1",
    "value": Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
  },
  Object {
    "op": "add",
    "path": "/users/2",
    "value": Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  },
  Object {
    "op": "add",
    "path": "/filter/0",
    "value": "FY",
  },
]
`;

exports[`can get users filtered by class and sorted 1`] = `
Object {
  "filter": Array [
    "FY",
  ],
  "sort": "name",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
    Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
    Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  ],
}
`;

exports[`can get users filtered by class and sorted 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/0",
    "value": Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
  },
  Object {
    "op": "add",
    "path": "/users/1",
    "value": Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
  },
  Object {
    "op": "add",
    "path": "/users/2",
    "value": Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  },
  Object {
    "op": "replace",
    "path": "/sort",
    "value": "name",
  },
  Object {
    "op": "add",
    "path": "/filter/0",
    "value": "FY",
  },
]
`;

exports[`can get users sorted by age 1`] = `
Object {
  "filter": Array [],
  "sort": "age",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
    Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
    Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  ],
}
`;

exports[`can get users sorted by age 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/0",
    "value": Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
  },
  Object {
    "op": "add",
    "path": "/users/1",
    "value": Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
  },
  Object {
    "op": "add",
    "path": "/users/2",
    "value": Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  },
  Object {
    "op": "replace",
    "path": "/sort",
    "value": "age",
  },
]
`;

exports[`can get users sorted by name 1`] = `
Object {
  "filter": Array [],
  "sort": "name",
  "users": Array [
    Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
    Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
    Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  ],
}
`;

exports[`can get users sorted by name 2`] = `
Array [
  Object {
    "op": "add",
    "path": "/users/0",
    "value": Object {
      "age": 18,
      "class": "FY",
      "name": "rachel",
    },
  },
  Object {
    "op": "add",
    "path": "/users/1",
    "value": Object {
      "age": 27,
      "class": "FY",
      "name": "joey",
    },
  },
  Object {
    "op": "add",
    "path": "/users/2",
    "value": Object {
      "age": 28,
      "class": "SY",
      "name": "chandler",
    },
  },
  Object {
    "op": "replace",
    "path": "/sort",
    "value": "name",
  },
]
`;
